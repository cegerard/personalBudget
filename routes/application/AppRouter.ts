import express from 'express';

import homeRoute from './home/routes';
import budgetsRoute from './budgets/routes';
import expensesRoute from './expenses/routes';
import balanceRoute from './balance/routes';
import ExpenseController from './expenses/controllers';
import BudgetController from './budgets/controllers';
import HomeController from './home/controllers';
import Authentication from '../../lib/security/authentication';
import { BalanceController } from './balance/controller';

class AppRouter {
  public router;

  constructor(
    homeController: HomeController,
    budgetController: BudgetController,
    expenseController: ExpenseController,
    balanceController: BalanceController,
  ) {
    this.router = express.Router();

    this.router.use('/', homeRoute.init(homeController));
    this.router.use(
      '/budgets',
      Authentication.ensureAuthenticated,
      budgetsRoute.init(budgetController)
    );
    this.router.use(
      '/expenses',
      Authentication.ensureAuthenticated,
      expensesRoute.init(expenseController)
    );
    this.router.use(
      '/balance',
      Authentication.ensureAuthenticated,
      balanceRoute.init(balanceController)
    );
  }
}

export default AppRouter;
