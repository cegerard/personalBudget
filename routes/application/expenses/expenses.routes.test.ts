import StatusCodes from 'http-status-codes';
import request from 'supertest';

import application from '../../../app';
import BudgetRepositoryStub from '../../../test/stubs/BudgetRepositoryStub';
import ExpenseRepositoryStub from '../../../test/stubs/ExpenseRepositoryStub';
import { authenticate } from '../../../test/authHelper';

const budgetRepository = application.budgetRepository as BudgetRepositoryStub;
const expenseRepository = application.expenseRepository as ExpenseRepositoryStub;

let app: any = null;

describe('Expenses with authentication', () => {
  beforeEach(async () => {
    budgetRepository.resetStore();
    expenseRepository.resetStore();
    app = await authenticate(application.app);
  });

  describe('GET /expenses', () => {
    it('should response with 200', async () => {
      await app.get('/expenses').expect(StatusCodes.OK);
    });

    it('should render expenses page', async () => {
      await app.get('/expenses').then((res: any) => {
        expect(res.text).toMatchSnapshot();
      });
    });
  });

  describe('GET /expenses/filter', () => {
    it('should response with 200', async () => {
      await app.get('/expenses/filter?budgetName=Essence').expect(StatusCodes.OK);
    });

    it('should render expenses filtered page', async () => {
      await app.get('/expenses/filter?budgetName=Essence').then((res: any) => {
        expect(res.text).toMatchSnapshot();
      });
    });
  });

  describe('POST /expenses', () => {
    const newExpense = {
      name: 'test expense',
      amount: 12,
      date: '2020-09-10',
      budgetlineId: '1',
    };

    it('should create a new expense', async () => {
      await app
        .post('/expenses')
        .set('Content-Type', 'application/json')
        .send(newExpense)
        .expect(StatusCodes.OK)
        .then(async () => {
          const expenses = await expenseRepository.find();
          const expensesFound = expenses.filter((expense: any) => {
            return expense.name === newExpense.name;
          });
          expect(expensesFound.length).toEqual(1);
          expect(expensesFound[0]).toMatchObject({
            _id: expect.any(String),
            name: newExpense.name,
            amount: newExpense.amount,
            date: newExpense.date,
            budgetLine: {
              _id: newExpense.budgetlineId,
              name: 'ElectricitÃ©',
            },
            owner: {
              id: '0001',
              name: 'admin istrator',
            },
          });
        });
    });

    // TODO add error mangement
    // it('should return 404 if budget not found', async () => {
    //   newExpense.budgetlineId = 'notfound';
    //   await request(app)
    //     .post('/expenses')
    //     .set('Content-Type', 'application/json')
    //     .send(newExpense)
    //     .expect(404);
    // });
  });

  describe('DELETE /expenses', () => {
    it('should remove expense from expense list and the budget line', async () => {
      await app
        .delete('/expenses/100')
        .expect(StatusCodes.NO_CONTENT)
        .then(async () => {
          const allExpenses = await expenseRepository.find(undefined);
          const expenseNotFound = allExpenses.find((expense: any) => {
            return expense._id === 100;
          });

          expect(expenseNotFound).toBeUndefined();

          const budgetList = await budgetRepository.getById('4');
          const expenseNotFoundInBudgetLine = budgetList.expenses.find((expense: any) => {
            return expense._id === 100;
          });

          expect(expenseNotFoundInBudgetLine).toBeUndefined();
        });
    });

    it('should return a 404 error when the expense can not be deleted', async () => {
      await app.delete('/expenses/404').expect(StatusCodes.NOT_FOUND);
    });

    it('should return a 500 error when the budget line does not exists', async () => {
      const newExpense = await expenseRepository.create({
        name: 'todelete',
        amount: 42,
        date: '',
        budgetLine: { _id: 'notExist', name: 'test' },
        owner: { id: '0001', name: 'admin istrator' },
      });

      await app.delete(`/expenses/${newExpense._id}`).expect(StatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return a 204 when expense does not exist in budget line', async () => {
      const newExpense = await expenseRepository.create({
        name: 'todelete',
        amount: 42,
        date: '',
        budgetLine: { _id: '1', name: 'test' },
        owner: { id: '0001', name: 'admin istrator' },
      });

      await app.delete(`/expenses/${newExpense._id}`).expect(StatusCodes.NO_CONTENT);
    });
  });

  describe('PATCH /expenses', () => {
    const expenseId = '100';

    let expenseBeforeUpdate: any;

    beforeEach(async () => {
      const expenses = await expenseRepository.find({ _id: expenseId });
      expenseBeforeUpdate = expenses[0];
    });

    it.each([
      ['name', 'new name'],
      ['amount', 42],
      ['date', '2021-02-02'],
    ])('should update the expense %s', async (field, value) => {
      await app
        .post(`/expenses/${expenseId}`)
        .set('Content-Type', 'application/json')
        .send({
          [field]: value,
          budgetLine: {
            _id: expenseBeforeUpdate.budgetLine._id,
          },
        })
        .expect(StatusCodes.OK);

      const expenses: any = await expenseRepository.find({ _id: expenseId });
      expect(expenses[0][field]).toEqual(value);
    });

    it('should not update the expense budget line ', async () => {
      await app
        .post(`/expenses/${expenseId}`)
        .set('Content-Type', 'application/json')
        .send({
          budgetLine: {
            _id: 'new_budget_line_id',
          },
        })
        .expect(StatusCodes.OK);

      const expenses = await expenseRepository.find({ _id: expenseId });
      expect(expenses[0]).toEqual(expenseBeforeUpdate);
    });

    it('should not add attributes not define in schema', async () => {
      await app
        .post(`/expenses/${expenseId}`)
        .set('Content-Type', 'application/json')
        .send({
          name: 'new name',
          notexist: 'should not add',
          dont: 'aie',
          budgetlineId: expenseBeforeUpdate.budgetLine._id,
        })
        .expect(StatusCodes.OK);

      const expenses: any = await expenseRepository.find({ _id: expenseId });
      expect(expenses[0].notexist).toBeUndefined();
      expect(expenses[0].dont).toBeUndefined();
    });

    it('should return 404 when expense does not exists', async () => {
      await app
        .post(`/expenses/not-exists`)
        .set('Content-Type', 'application/json')
        .send({ amount: 100 })
        .expect(StatusCodes.NOT_FOUND);
    });

    it('should render expenses page', async () => {
      await app
        .post(`/expenses/${expenseId}`)
        .set('Content-Type', 'application/json')
        .send({ amount: 100, budgetlineId: expenseBeforeUpdate.budgetLine._id })
        .expect(StatusCodes.OK)
        .then((res: any) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });
});

describe('Expenses without authentication', () => {
  beforeEach(() => {
    app = request(application.app);
  });

  describe('GET /expenses', () => {
    it('should response with 401', async () => {
      await app.get('/expenses').expect(StatusCodes.MOVED_TEMPORARILY);
    });

    it('should render home page', async () => {
      await app.get('/expenses').then((res: any) => {
        expect(res.text).toMatchSnapshot();
      });
    });
  });

  describe('GET /expenses/filter', () => {
    it('should response with 401', async () => {
      await app.get('/expenses/filter?budgetName=Essence').expect(StatusCodes.MOVED_TEMPORARILY);
    });

    it('should render the home page', async () => {
      await app.get('/expenses/filter?budgetName=Essence').then((res: any) => {
        expect(res.text).toMatchSnapshot();
      });
    });
  });

  describe('POST /expenses', () => {
    const newExpense = {
      name: 'test expense',
      amount: 12,
      date: '2020-09-10',
      budgetlineId: '1',
    };

    it('should do nothing', async () => {
      await app
        .post('/expenses')
        .set('Content-Type', 'application/json')
        .send(newExpense)
        .expect(StatusCodes.MOVED_TEMPORARILY)
        .then(async () => {
          const expenses = await expenseRepository.find(undefined);
          const expensesFound = expenses.filter((expense: any) => {
            return expense.name === newExpense.name;
          });
          expect(expensesFound.length).toEqual(0);
        });
    });
  });

  describe('DELETE /expenses', () => {
    it('should return unauthorized', async () => {
      await app.delete('/expenses/100').expect(StatusCodes.MOVED_TEMPORARILY);
    });

    it('should return a 404 error when the expense can not be deleted', async () => {
      await app.delete('/expenses/404').expect(StatusCodes.MOVED_TEMPORARILY);
    });
  });

  describe('PATCH /expenses', () => {
    const expenseId = '100';

    let expenseBeforeUpdate: any;

    beforeEach(async () => {
      const expenses = await expenseRepository.find({ _id: expenseId });
      expenseBeforeUpdate = expenses[0];
    });

    it('should return unauthorized', async () => {
      await app
        .post(`/expenses/${expenseId}`)
        .set('Content-Type', 'application/json')
        .send({
          budgetLine: {
            _id: 'new_budget_line_id',
          },
        })
        .expect(StatusCodes.MOVED_TEMPORARILY);
    });
  });
});
