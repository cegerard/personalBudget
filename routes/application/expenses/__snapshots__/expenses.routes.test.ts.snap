// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expenses with authentication GET /expenses should render expenses page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div class=\\"row\\"><h1 class=\\"mb-4 mt-3 col-md-9\\">Expenses<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addExpenseModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1><div class=\\"col-md-3 mb-4 mt-3\\"><div class=\\"input-group\\" style=\\"padding-top: 4%\\"><input class=\\"form-control\\" id=\\"expense-filter\\" type=\\"text\\" placeholder=\\"Budget Name\\" aria-label=\\"Budget Name\\" aria-describedby=\\"budget-name-filter\\"><div class=\\"input-group-append\\"><button class=\\"btn btn-outline-secondary\\" id=\\"budget-name-filter\\" type=\\"button\\" onclick=\\"filterExpenseOnBudgetName()\\">Filter</button></div></div></div></div><table class=\\"table\\"><thead><tr><th>Expense</th><th>Amount</th><th>Budget Line</th><th>Date</th><th>Action</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>Essence</td><td>2020-08-29</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;100&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;100&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>Ka+</td><td>30 €</td><td>Essence</td><td>2020-08-22</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;101&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;101&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>courses</td><td>61.45 €</td><td>Course</td><td>2020-10-08</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;200&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;200&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr></table><div class=\\"modal fade\\" id=\\"addExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Add Expense</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/expenses\\"><div class=\\"form-group\\"><label for=\\"expenseName\\">Name</label><input class=\\"form-control\\" id=\\"expenseName\\" name=\\"name\\" type=\\"text\\" required><label for=\\"expenseAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelp\\">Expense amount must be gretter than 0</small><label for=\\"expenseDate\\">Date</label><input class=\\"form-control\\" id=\\"expenseDate\\" name=\\"date\\" type=\\"date\\" required><label for=\\"budgetLine\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLine\\" name=\\"budgetlineId\\"><option value=\\"1\\">Electricité</option><option value=\\"2\\">Crédit voiture</option><option value=\\"3\\">Crédit immobilier</option><option value=\\"4\\">Essence</option><option value=\\"5\\">Assurance voiture</option><option value=\\"6\\">Alimentation</option><option value=\\"7\\">Animaux</option><option value=\\"8\\">Enfants</option></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div><div class=\\"modal fade\\" id=\\"editExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Edit Expense</h4></div><div class=\\"modal-body\\"><form id=\\"editExpenseForm\\" method=\\"POST\\"><div class=\\"form-group\\"><label for=\\"expenseNameEdit\\">Name</label><input class=\\"form-control\\" id=\\"expenseNameEdit\\" name=\\"name\\" type=\\"text\\"><label for=\\"expenseAmountEdit\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmountEdit\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\"><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelpEdit\\">Expense amount must be gretter than 0</small><label for=\\"expenseDateEdit\\">Date</label><input class=\\"form-control\\" id=\\"expenseDateEdit\\" name=\\"date\\" type=\\"date\\"><label for=\\"budgetLineEdit\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLineEdit\\" name=\\"budgetlineId\\" disabled></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Edit</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Expenses with authentication GET /expenses/filter should render expenses filtered page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div class=\\"row\\"><h1 class=\\"mb-4 mt-3 col-md-9\\">Expenses<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addExpenseModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1><div class=\\"col-md-3 mb-4 mt-3\\"><div class=\\"input-group\\" style=\\"padding-top: 4%\\"><input class=\\"form-control\\" id=\\"expense-filter\\" type=\\"text\\" placeholder=\\"Budget Name\\" aria-label=\\"Budget Name\\" aria-describedby=\\"budget-name-filter\\"><div class=\\"input-group-append\\"><button class=\\"btn btn-outline-secondary\\" id=\\"budget-name-filter\\" type=\\"button\\" onclick=\\"filterExpenseOnBudgetName()\\">Filter</button></div></div></div></div><table class=\\"table\\"><thead><tr><th>Expense</th><th>Amount</th><th>Budget Line</th><th>Date</th><th>Action</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>Essence</td><td>2020-08-29</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;100&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;100&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>Ka+</td><td>30 €</td><td>Essence</td><td>2020-08-22</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;101&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;101&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr></table><div class=\\"modal fade\\" id=\\"addExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Add Expense</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/expenses\\"><div class=\\"form-group\\"><label for=\\"expenseName\\">Name</label><input class=\\"form-control\\" id=\\"expenseName\\" name=\\"name\\" type=\\"text\\" required><label for=\\"expenseAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelp\\">Expense amount must be gretter than 0</small><label for=\\"expenseDate\\">Date</label><input class=\\"form-control\\" id=\\"expenseDate\\" name=\\"date\\" type=\\"date\\" required><label for=\\"budgetLine\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLine\\" name=\\"budgetlineId\\"><option value=\\"1\\">Electricité</option><option value=\\"2\\">Crédit voiture</option><option value=\\"3\\">Crédit immobilier</option><option value=\\"4\\">Essence</option><option value=\\"5\\">Assurance voiture</option><option value=\\"6\\">Alimentation</option><option value=\\"7\\">Animaux</option><option value=\\"8\\">Enfants</option></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div><div class=\\"modal fade\\" id=\\"editExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Edit Expense</h4></div><div class=\\"modal-body\\"><form id=\\"editExpenseForm\\" method=\\"POST\\"><div class=\\"form-group\\"><label for=\\"expenseNameEdit\\">Name</label><input class=\\"form-control\\" id=\\"expenseNameEdit\\" name=\\"name\\" type=\\"text\\"><label for=\\"expenseAmountEdit\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmountEdit\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\"><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelpEdit\\">Expense amount must be gretter than 0</small><label for=\\"expenseDateEdit\\">Date</label><input class=\\"form-control\\" id=\\"expenseDateEdit\\" name=\\"date\\" type=\\"date\\"><label for=\\"budgetLineEdit\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLineEdit\\" name=\\"budgetlineId\\" disabled></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Edit</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Expenses with authentication PATCH /expenses should render expenses page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div class=\\"row\\"><h1 class=\\"mb-4 mt-3 col-md-9\\">Expenses<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addExpenseModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1><div class=\\"col-md-3 mb-4 mt-3\\"><div class=\\"input-group\\" style=\\"padding-top: 4%\\"><input class=\\"form-control\\" id=\\"expense-filter\\" type=\\"text\\" placeholder=\\"Budget Name\\" aria-label=\\"Budget Name\\" aria-describedby=\\"budget-name-filter\\"><div class=\\"input-group-append\\"><button class=\\"btn btn-outline-secondary\\" id=\\"budget-name-filter\\" type=\\"button\\" onclick=\\"filterExpenseOnBudgetName()\\">Filter</button></div></div></div></div><table class=\\"table\\"><thead><tr><th>Expense</th><th>Amount</th><th>Budget Line</th><th>Date</th><th>Action</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>100 €</td><td>Essence</td><td>2020-08-29</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;100&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;100&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>Ka+</td><td>30 €</td><td>Essence</td><td>2020-08-22</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;101&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;101&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>courses</td><td>61.45 €</td><td>Course</td><td>2020-10-08</td><td> <button class=\\"btn btn-primary btn-sm mr-1\\" onclick=\\"editExpense(&quot;200&quot;)\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;200&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr></table><div class=\\"modal fade\\" id=\\"addExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Add Expense</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/expenses\\"><div class=\\"form-group\\"><label for=\\"expenseName\\">Name</label><input class=\\"form-control\\" id=\\"expenseName\\" name=\\"name\\" type=\\"text\\" required><label for=\\"expenseAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelp\\">Expense amount must be gretter than 0</small><label for=\\"expenseDate\\">Date</label><input class=\\"form-control\\" id=\\"expenseDate\\" name=\\"date\\" type=\\"date\\" required><label for=\\"budgetLine\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLine\\" name=\\"budgetlineId\\"><option value=\\"1\\">Electricité</option><option value=\\"2\\">Crédit voiture</option><option value=\\"3\\">Crédit immobilier</option><option value=\\"4\\">Essence</option><option value=\\"5\\">Assurance voiture</option><option value=\\"6\\">Alimentation</option><option value=\\"7\\">Animaux</option><option value=\\"8\\">Enfants</option></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div><div class=\\"modal fade\\" id=\\"editExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Edit Expense</h4></div><div class=\\"modal-body\\"><form id=\\"editExpenseForm\\" method=\\"POST\\"><div class=\\"form-group\\"><label for=\\"expenseNameEdit\\">Name</label><input class=\\"form-control\\" id=\\"expenseNameEdit\\" name=\\"name\\" type=\\"text\\"><label for=\\"expenseAmountEdit\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmountEdit\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\"><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelpEdit\\">Expense amount must be gretter than 0</small><label for=\\"expenseDateEdit\\">Date</label><input class=\\"form-control\\" id=\\"expenseDateEdit\\" name=\\"date\\" type=\\"date\\"><label for=\\"budgetLineEdit\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLineEdit\\" name=\\"budgetlineId\\" disabled></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Edit</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Expenses without authentication GET /expenses should render home page 1`] = `"Unauthorized. Redirecting to /"`;

exports[`Expenses without authentication GET /expenses/filter should render the home page 1`] = `"Unauthorized. Redirecting to /"`;
